src/compiler/checker.ts

```diff
@@ -39007,8 +39007,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         // Get accessors without matching set accessors
         // Enum members
         // Object.defineProperty assignments with writable false or no setter
-        // Unions and intersections of the above (unions and intersections eagerly set isReadonly on creation)
-        return !!(getCheckFlags(symbol) & CheckFlags.Readonly ||
+        // Unions and intersections of the above
+        const checkFlags = getCheckFlags(symbol);
+        if (checkFlags & CheckFlags.Synthetic) {
+            // unions and intersections eagerly compute Readonly flag on creation
+            return !!(checkFlags & CheckFlags.Readonly);
+        }

```

Comment:
1. The issue arised because `getDeclarationModifierFlagsFromSymbol(symbol)` was returning `ModifierFlags.Readonly` based on the `.valueDeclaration`.
2. `.valueDeclaration` is only set on union/intersection properties when it's "uniform"(ish). In this example, one source property symbol had a `.valueDeclaration` and the other one hadn't (the one from the mapped type hadn't it)
3. but given that, as the existing comment states, the readonly flag is precomputed eagerly for union/intersections - this is really the only thing the compiler has to check in that situation. So I leveraged that for the fix here.